File: Torch/ConcatTable.lua
-- Copyright 2016-present, Facebook, Inc.
-- All rights reserved.
--
-- This source code is licensed under the license found in the
-- LICENSE file in the root directory of this source tree.

return function(sparseconvnet)
  local ConcatTable, parent = torch.class(
    'sparseconvnet.ConcatTable', 'nn.ConcatTable', sparseconvnet)

  function ConcatTable:__init()
    parent.__init(self)
    self.modules={}
    self.output={}
    self.gradInput={
      features=torch.Tensor()
    }
    sparseconvnet.shareShared(self)
  end

  function ConcatTable:add(module)
    table.insert(self.modules,module)
    sparseconvnet.shareShared(self)
    return self
  end

  function ConcatTable:updateOutput(input)
    for i = 1,#self.modules do
      self.output[i]=self.modules[i]:forward(input)
    end
    for i = #self.modules+1,#self.output do
      self.output[i]=nil
    end
    return self.output
  end

  function ConcatTable:backward(input, gradOutput)
    local gradInputs={}
    for i = 1,#self.modules do
      gradInputs[i]=self.modules[i]:backward(input,gradOutput[i],scale)
    end
    self.gradInput.features:resizeAs(
      gradInputs[1].features):copy(gradInputs[1].features)
    for i=2,#self.modules do
      self.gradInput.features:add(gradInputs[i].features)
    end
    return self.gradInput
  end

  function ConcatTable:clearState()
    for _,m in ipairs(self.modules) do
      m:clearState()
    end
    self.output={}
    self.gradInput={features=self.gradInput.features:set()}
  end

  function ConcatTable:suggestInputSize(nOut)
    return self.modules[1]:suggestInputSize(nOut)
  end
end

File: PyTorch/sparseconvnet/legacy/concatTable.py
# Copyright 2016-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import torch
import sparseconvnet
from torch.legacy.nn import ConcatTable as C
from ..utils import toLongTensor, dim_typed_fn, optionalTensor, nullptr, set
from .sparseConvNetTensor import SparseConvNetTensor


class ConcatTable(C):
    def __init__(self):
        C.__init__(self)
        self.gradInput = torch.Tensor()

    def updateOutput(self, input):
        self.output = []
        for m in self.modules:
            self.output.append(m.forward(input))
        return self.output

    def backward(self, input, gradOutput, scale=1):
        self.gradInput.resize_as_(input.features).zero_()
        for m, g in zip(self.modules, gradOutput):
            self.gradInput.add_(m.backward(input, g, scale))
        return self.gradInput

    def clearState(self):
        self.output = None
        set(self.gradInput)
        for m in self.modules:
            m.clearState()

    def suggestInputSize(self, nOut):
        return self.modules[0].suggestInputSize(nOut)
